cajita/src/Cajita_ParticleGridDistributor.hpp

```diff
@@ -0,0 +1,390 @@
+                         positions( p, d ) > local_high[d] )
+                    {
+                        result += 1;
+                        break;
+                    }
+            },
+            comm_count );
+
+        MPI_Allreduce( MPI_IN_PLACE, &comm_count, 1, MPI_INT, MPI_SUM,
+                       _local_grid.globalGrid().comm() );
+
+        return comm_count;
+    }
+
+    // Build neighbor topology of 27 nearest 3D neighbors. Some of the ranks in
+    // this list may be invalid.
+    template <std::size_t NSD = num_space_dim>
+    std::enable_if_t<3 == NSD, std::vector<int>> getTopology()
+    {
+        std::vector<int> topology( 27, -1 );
+        int nr = 0;
+        for ( int k = -1; k < 2; ++k )
+            for ( int j = -1; j < 2; ++j )
+                for ( int i = -1; i < 2; ++i, ++nr )
+                    topology[nr] = _local_grid.neighborRank( i, j, k );
+        return topology;
+    }
+
+    // Build neighbor topology of 8 nearest 2D neighbors. Some of the ranks in
+    // this list may be invalid.
+    template <std::size_t NSD = num_space_dim>
+    std::enable_if_t<2 == NSD, std::vector<int>> getTopology()
+    {
+        std::vector<int> topology( 9, -1 );
+        int nr = 0;
+        for ( int j = -1; j < 2; ++j )
+            for ( int i = -1; i < 2; ++i, ++nr )
+                topology[nr] = _local_grid.neighborRank( i, j );
+        return topology;
+    }
+
+    // Locate the particles in the local grid and get their destination rank.
+    // Particles are assumed to only migrate to a location in the nearest
+    // neighbor halo or stay on this rank. If the particle crosses a global
+    // periodic boundary, wrap it's coordinates back into the domain.
+    template <class PositionSliceType, class NeighborRankView,
+              class DestinationRankView>
+    void getMigrateDestinations( const NeighborRankView& neighbor_ranks,
+                                 DestinationRankView& destinations,
+                                 PositionSliceType& positions )
+    {
+        using execution_space = typename PositionSliceType::execution_space;
+
+        // Check within the local domain.
+        const auto& local_mesh =
+            Cajita::createLocalMesh<Kokkos::HostSpace>( _local_grid );
+
+        // Use global domain for periodicity.
+        const auto& global_grid = _local_grid.globalGrid();
+        const auto& global_mesh = global_grid.globalMesh();
+
+        Kokkos::Array<double, num_space_dim> local_low{};
+        Kokkos::Array<double, num_space_dim> local_high{};
+        Kokkos::Array<bool, num_space_dim> periodic{};
+        Kokkos::Array<double, num_space_dim> global_low{};
+        Kokkos::Array<double, num_space_dim> global_high{};
+        Kokkos::Array<double, num_space_dim> global_extent{};
+
+        for ( std::size_t d = 0; d < num_space_dim; ++d )
+        {
+            local_low[d] = local_mesh.lowCorner( Cajita::Own(), d );
```

Comment:
We do this in a lot of places. No need to add in this PR but it would be nice to have functions for these that do not take a dimension argument and just return a `Kokkos::Array` for all dimensions.